### ISSUE TITLE
remove sw-specific types from Python API that can simply be numpy arrays or eigen matrices

### ISSUE THREAD

**Issue by sheryjoe** [SCIInstitute/ShapeWorks#830]
shapeworks vector class should have an API that returns a numpy array. this will enable numpy operations such as min, max .. etc

**Comment by cchriste** [comment]
There will soon be no shapeworks.Vector class and instead all relevant Python funtions will accept/return numpy arrays (e.g., Vector and Point), standard Python buffers (e.g., Coord and Dims), or Eigen matrices. 

This will make using our API much more natural in Python as well as more compatible with any other Python libraries. It will also enable passing buffer data without requiring a copy (*we should create a new GitHub issue for this wrt itk Images in particular). @archanasri and I created examples of this working for a few functions and the Image constructor the day before the Thanksgiving holiday and we are fleshing it out now. 

A similar functionality using std::vectors and Eigen matrices would be desirable on the C++ side, thereby reducing our nearly overt dependence on ITK, but the same code (using py::buffer_t<type>) would be required on the Python side either way, so this is the optimal place to start.

Here are a couple illustrations:

**Comment by sheryjoe** [comment]
@cchriste is this pushed to pybind branch?

**Comment by sheryjoe** [comment]
> @cchriste is this pushed to pybind branch?

aha, pybind_fix - I will branch out of this branch then.

**Comment by cchriste** [comment]
This is really preliminary with only an example constructor and a couple functions showing how to accept or return a std::vector (automatically converted by pybind11). For the constructor, it doesn't know the origin, spacing or coordsys, so we'll need to create Image::setSpacing(vector) function. Then it can be used something like this:
```
img = sw.Image(nparr)
img.setOrigin(o)
img.setSpacing([1, 2.6, 0.667])
```

The toArray function in ShapeworksPython.cpp shows how to return a numpy array, but it copies data. 
There are ways to do this that share memory with its accompanying advantages. Here are a couple useful pages for this:
https://pybind11.readthedocs.io/en/stable/advanced/pycpp/numpy.html#buffer-protocol
https://stackoverflow.com/questions/44659924/returning-numpy-arrays-via-pybind11 (the answer here is a good example)

We really don't need Coord, Dims, Point, Vector, Matrix, Matrix44, IPoint3, FPoint3, TransformPtr, or Axis. Most of these can  just be passed back and forth as py::buffers or py::arrays (often simply couched as std::vectors) or Eigen matrices. 
https://pybind11.readthedocs.io/en/stable/advanced/cast/eigen.html

**Comment by cchriste** [comment]
Reflecting on the current C++ Image class interface, I wonder if we might simply send and receive Eigen matrices and std::vectors rather than burden the user with a zillion different ways of passing three, nine, or 16 numbers around. They can be casted and converted to the required ITK classes internally. Taking this pass at the Python interface will essentially do this. For example, the Image::logicalToPhysical function looks like this (docstrings are missing to keep this simple):
```
  .def("logicalToPhysical", [](Image& image, std::vector<long>& c) { return image.logicalToPhysical(Coord({c[0], c[1], c[2]})); })
```
but if that function's signature in Image simply took a std::vector<long> the Python binding could simply be this:
```
  .def("logicalToPhysical", &Image::logicalToPhysical)
```
The conversion to Coord would then be done in the Image function.

**Comment by sheryjoe** [comment]
Are we planning to demonstrate this in the users meeting? or should I stick to the pybind branch for the tutorial notebooks for the meeting next week?

**Comment by archanasri** [comment]
Done with commit f30c2a0

**Comment by cchriste** [comment]
This is great! The only thing I'd like to add to this is a generalization to Matrices and a removal of all unnecessary classes from the Python bindings. This will make it so things like Image.center don't return some proprietary shapeworks.point but instead return a numpy array. In the bindings code we can just do this with a std::vector.

For the Matrices we can create a separate GitHub issue if you want (or maybe there already is) and use Eigen since there is nice documentation for this (next tab down from the numpy docs in pybind11). The advantage is users will be able to manipulate transforms using the comprehensive Eigen::Matrix class which already has all the Python bindings in place. Handy for Image, but huge for classes down the road that actually use big matrices. 

To summarize, I'm reopening for two reasons:
1) to remove shapeworks-specific array classes, which involves just a tiny bit more work than the previous commit.
2) to use Eigen::Matrix for taking and returning matrices (it uses numpy under the hood; see [pybind Eigen docs](https://pybind11.readthedocs.io/en/stable/advanced/cast/eigen.html)

**Comment by cchriste** [comment]
Here's a reference to help with Eigen for transform matrices: https://pybind11.readthedocs.io/en/latest/advanced/cast/eigen.html

**Comment by cchriste** [comment]
In case anyone was wondering, accepting a std::vector for a 3-item array is faster than accepting a py::array. Here are the results:
Passing a py::array
  // passing list (i.e., [1,2,3)]: 10K runs in 0.4s, 100K in 3.9s, 1M in **39.6129**
  // passing ndarray: 10K runs in ?s, 100K in 3.5s, 1M in **35.5292**
  .def("setSpacing",
       [](Image& self, const py::array_t<double>& v) -> decltype(auto) {
         auto r = v.unchecked<1>();
         return self.setSpacing(makeVector({r(0), r(1), r(2)}));
       },
       "set image spacing, the size of each pixel",
       "spacing"_a=py::array_t<double>({1.0, 1.0, 1.0}))
Passing a std::vector… the winner!!
  // passing list (i.e., [1,2,3)]: 100K in 3.33s, 1M in 33.4944
  // passing ndarray: 100K in 3.4s, 1M in 33.9584
  .def("setSpacing",
       [](Image& self, std::vector<double>& v) -> decltype(auto) {
         return self.setSpacing(makeVector({v[0], v[1], v[2]}));
       },
       "set image spacing, the size of each pixel",
       "spacing"_a=makeVector({1.0, 1.0, 1.0}))

**Comment by cchriste** [comment]
Related to #903

**Comment by cchriste** [comment]
This is finished in the pybind branch and ready for QA.

**Comment by archanasri** [comment]
@sheryjoe yes it is with the pybind PR
